User Record Import Updates

dtgreiner opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue)
[//]: # (List any new dependencies or relevant ADRs)

The User records were not being updated during the import process. This update allows them to be updated but also keeps it so they cannot be deleted if not included in the import.

## Type of change
[//]: # (e.g., Bug fix, New feature, Documentation, Code clean-up, Dependency update)
Bug fix

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] Any major architectural changes are supported by an approved ADR (Architectural Decision Record) 
- [ ] I have updated the documentation (or not applicable)
- [X] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected

dtgreiner committed:
Allow user records in the import process to be updated

dtgreiner committed:
remove test that should not have been included in the last  commit

eanders commented on drivers/hmis_csv_twenty_twenty_four/app/models/hmis_csv_twenty_twenty_four/importer/user.rb:
```diff
@@ -22,8 +22,14 @@ def self.warehouse_class
+    # end
+
+    # # Don't ever mark these for deletion
+    def self.prevent_import_deletions?
+      true
```
Can you check Project and Organization as well, I believe they do not allow deletions, but maybe through a different mechanism.  Perhaps we can update them to use this?

dtgreiner replied:
yeah, I can give them a look. My gut says that they were doing the same thing as the User records and as a result are probably not allowing updates too.

eanders commented on drivers/hmis_csv_importer/spec/models/importer/twenty_twenty_four/base_spec.rb:
```diff
@@ -32,6 +32,60 @@
+      expect(GrdaWarehouse::Hud::User.count).to eq(4)
+    end
+
+    it 'user emails match expected' do
+      user_emails = GrdaWarehouse::Hud::User.all.pluck(:UserEmail).sort
+      expected_emails = [
+        'freddie@example.com',
+        'ryleigh@example.com',
+        'jackson@example.com',
+        'tamara@example.com',
+      ].sort
+      expect(user_emails).to eq(expected_emails)
+    end
+
+    describe 'after second import' do
+      before(:all) do
+        import_hmis_csv_fixture(
+          'drivers/hmis_csv_importer/spec/fixtures/files/twenty_twenty_four/user_test_2',
+          version: 'AutoMigrate',
+          run_jobs: true,
+        )
+      end
+
+      it 'users not in import file are not deleted' do
+        # We still expect 4 users even through the new import file only has 3 users included.
+        # Users should not be deleted.
+        expect(GrdaWarehouse::Hud::User.count).to eq(4)
+      end
+
+      it 'users emails are updated to match import file' do
+        user_emails = GrdaWarehouse::Hud::User.all.pluck(:UserEmail).sort
+        expected_emails = [
+          'freddie+edited@example.com',
+          'ryleigh+edited@example.com',
+          'jackson@example.com',
+          'tamara@example.com',
+        ].sort
+        expect(user_emails).to eq(expected_emails)
+      end
+    end
```
This is great, thank you!

eanders reviewed (approved):
Before we merge, can you just take a look at if we might have a similar problem with Projects and Organizations, and if we should update them to work in the same way?

Otherwise, looks great as a solution to the user update issue.

dtgreiner committed:
clean up for project and organizations & add more tests

dtgreiner commented:
@eanders I've added in the changes for Organization & Project. Project is inherently not allowing deletions due to how it is being to identify involved records (see [here](https://github.com/greenriver/hmis-warehouse/blob/620f12d04917b1914948e4d275a77238128f35fd/drivers/hmis_csv_importer/app/models/hmis_csv_importer/importer/importer.rb#L488) and subsequently [here](https://github.com/greenriver/hmis-warehouse/blob/620f12d04917b1914948e4d275a77238128f35fd/drivers/hmis_csv_twenty_twenty_four/app/models/hmis_csv_twenty_twenty_four/importer/project.rb#L28)) but I've included the change to that class to make it easier to identify which file types are not allowing deletions. Also, I've cleaned up and split out tests for each of these files on import.

dtgreiner committed:
Merge branch 'release-150' into dg/import_user_change-7064

eanders commented on drivers/hmis_csv_importer/spec/models/importer/twenty_twenty_four/base_spec.rb:
```diff
@@ -32,6 +32,241 @@
+        )
+      end
+
+      it 'organizations not in import file are not deleted' do
```
```suggestion
      it 'organizations not in import file are deleted' do
```

eanders reviewed (approved):
Thanks @dtgreiner !  Looks good.  Just the one thought that the comment on a test is inverted.

eanders committed:
Merge branch 'release-150' into dg/import_user_change-7064

dtgreiner committed:
Update drivers/hmis_csv_importer/spec/models/importer/twenty_twenty_four/base_spec.rb

Co-authored-by: Elliot <elliot@greenriver.org>
